armi/plugins.py

```diff
@@ -600,10 +600,72 @@ def defineSystemBuilders() -> Dict[str, Callable[[str], "Composite"]]:
         """
 
 
-def getNewPluginManager() -> pluginManager.ArmiPluginManager:
+class UserPlugin(ArmiPlugin):
     """
-    Return a new plugin manager with all of the hookspecs pre-registered.
+    This is a variation on the usual ArmiPlugin that is meant to be read in during run time,
+    from a hard-coded path in a settings file, or the like.
+
+    This is obviously a more limited use-case than the usual ArmiPlugin, as those are meant
+    to be defined at import time, instead of run time. As such, this class has some built-in
+    tooling to limit how these run-time plugins are used. They are meant to be more limited.
+
+    NOTE: The usual ArmiPlugin is much more flexible, if the UserPlugin does not support what
+    you want to do, just use an ArmiPlugin.
     """
+
+    def __init__(self, *args, **kwargs):
+        ArmiPlugin.__init__(self, *args, **kwargs)
+        self.__enforceLimitations()
+
+    def __enforceLimitations(self):
+        """
+        This method is designed to enforce that UserPlugins are more limited
+        than regular ArmiPlugins.
+        UserPlugins are different from regular plugins in that they can be defined during
+        a run, and as such, we want to limit how flexible they are, so we can correctly
+        corral their side effects during a run.
+        """
+        if issubclass(self.__class__, UserPlugin):
+            assert len(self.__class__.defineSettings()) == 0, "TODO"
+            # NOTE: These are the class methods that we are staunchly _not_ allowing people
+            # to change in this class. If you need these, please use a regular ArmiPlugin.
+            self.defineParameterRenames = lambda: None
+            self.defineSettings = lambda: []
+            self.defineSettingsValidators = lambda: []
```

Comment:
can we raise errors if set rather than suppress silently?

Reply:
Well, a few lines up we DO raise an error if `defineSettings()` has anything in it.

But, from a strictly technological perspective, I'm not sure _how_ to detect code in the other two functions. You see, function content is mutable in Python (as you see here), so there is no guarantee that you can correctly determine if a method is empty.  (This is particularly annoying because "docstrings" are actually Python string objects, so technically any method with a docstring is not empty.

I can look at this again, but I don't know how to test these other two and be 100% sure I was correct.

Reply:
They all tend to evaluate to lists or dictionaries so I was thinking you could do it very simialry to how you dothe same as you do settings

assert  not self.__class__.defineSettings(), "defined settings are not supported for user plugins"
assert  not self.__class__.defineParameterRenames(), "param renames are not supported for user plugins"
ssert  not self.__class__.defineSettingsValidators(),  "settings validators are not supported for user plugins"

Reply:
that makes sense and defineSettings is the big bang for buck. if you cant define settings people will easily notice they cant define validators.